File: src/main/java/me/hhitt/disasters/listener/ItemPickupListener.java
package me.hhitt.disasters.listener;

import me.hhitt.disasters.arena.Arena;
import me.hhitt.disasters.arena.ArenaManager;
import me.hhitt.disasters.game.drop.ItemDropManager;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityPickupItemEvent;
import org.bukkit.event.entity.ItemMergeEvent;
import org.bukkit.persistence.PersistentDataType;

public final class ItemPickupListener implements Listener {

    private final ArenaManager arenaManager;

    public ItemPickupListener(final ArenaManager arenaManager) {
        this.arenaManager = arenaManager;
    }

    @EventHandler
    public void onItemPickup(final EntityPickupItemEvent event) {
        if (!(event.getEntity() instanceof Player)) {
            return;
        }
        final Player player = (Player) event.getEntity();
        final Arena arena = arenaManager.getArena(player);
        if (arena == null) {
            return;
        }

        if (event.getItem().getPersistentDataContainer().has(ItemDropManager.DROP_KEY, PersistentDataType.BYTE)) {
            ItemDropManager.onPickup(arena, event.getItem());
            return;
        }

        event.setCancelled(true);
    }

    @EventHandler
    public void onItemMerge(final ItemMergeEvent event) {
        if (event.getEntity().getPersistentDataContainer().has(ItemDropManager.DROP_KEY, PersistentDataType.BYTE)
            || event.getTarget().getPersistentDataContainer().has(ItemDropManager.DROP_KEY, PersistentDataType.BYTE)) {
            event.setCancelled(true);
        }
    }
}

File: src/main/kotlin/me/hhitt/disasters/listener/ItemPickupListener.kt
package me.hhitt.disasters.listener

import me.hhitt.disasters.arena.ArenaManager
import me.hhitt.disasters.game.drop.ItemDropManager
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityPickupItemEvent
import org.bukkit.event.entity.ItemMergeEvent
import org.bukkit.persistence.PersistentDataType

class ItemPickupListener(private val arenaManager: ArenaManager) : Listener {

    @EventHandler
    fun onItemPickup(event: EntityPickupItemEvent) {
        val player = event.entity as? Player ?: return
        val arena = arenaManager.getArena(player) ?: return

        // Allow picking up disaster drop items
        val itemEntity = event.item
        if (itemEntity.persistentDataContainer.has(ItemDropManager.DROP_KEY, PersistentDataType.BYTE)) {
            ItemDropManager.onPickup(arena, itemEntity)
            return
        }

        event.isCancelled = true
    }

    @EventHandler
    fun onItemMerge(event: ItemMergeEvent) {
        // Prevent disaster drop items from merging with other items
        if (event.entity.persistentDataContainer.has(ItemDropManager.DROP_KEY, PersistentDataType.BYTE) ||
            event.target.persistentDataContainer.has(ItemDropManager.DROP_KEY, PersistentDataType.BYTE)) {
            event.isCancelled = true
        }
    }
}
